### PR TITLE
add TestData::getTestSamples()

### PR DESCRIPTION
**Pull request by andreasBihlmaier**
Documentation for cv::ml::TrainData::setTrainTestSplitRatio explicitly states "Please, note that for each of TrainData::getTrain* there is corresponding TrainData::getTest*, so that the test subset can be retrieved and processed as well."

For some reason getTestSamples() was missing.

### CODE REVIEW

File: modules/ml/include/opencv2/ml.hpp [hunk 1]
```diff
@@ -181,6 +181,7 @@ class CV_EXPORTS_W TrainData
     TrainData::getClassLabels.
      */
     CV_WRAP virtual Mat getTrainNormCatResponses() const = 0;
+    CV_WRAP virtual Mat getTestSamples() const = 0;
```
Comment (mshabunin) [inline modules/ml/include/opencv2/ml.hpp]: Try to remove the `virtual` specifier to avoid ABI compatibility breakage.

**Comment by andreasBihlmaier** [comment]
@mshabunin I agree that adding the function breaks ABI.
But I'd argue that the function is really missing from the 'interface' definition and thus ABI should be broken. Is there a "3.1" branch, which allows ABI/API changing pull-requests?

**Comment by mshabunin** [comment]
@andreasBihlmaier , it is assumed that version 3.1 will be binary compatible with 3.0. Actually I don't see why this function should be _virtual_. If you replace `samples` with `getSamples` in function body, than it will be possible to change its behaviour by overriding `getSamples` and `getTestSampleIdx`.

**Comment by andreasBihlmaier** [comment]
The problem is that https://github.com/andreasBihlmaier/opencv/blob/master/modules/ml/include/opencv2/ml.hpp#L130 states that TrainData "only specifies the interface", i.e. the closest thing to that in C++ is a pure virtual class, which TrainData actually is at the moment - all existing functions are pure virtual and are implemented in TrainDataImpl. Why should that be different for the missing getTestSamples? Especially given that getTestSamples is also pure virtual in TrainData and implemented in TrainDataImpl3

**Comment by mshabunin** [comment]
@andreasBihlmaier , you can try to move new function to the end of the virtual function list and see whether the ABI will be compatible.

**Comment by andreasBihlmaier** [comment]
@mshabunin unfortunately, this does not suffice: http://pullrequest.opencv.org/buildbot/builders/precommit_linux64/builds/4958/steps/Compare%20ABI%20dumps/logs/report-html#Type_Binary_Problems_High

**Comment by mshabunin** [comment]
@andreasBihlmaier , I don't see other alternatives:

- the added method is not used (and probably will not be used) in the ML algorithms
- I can't imagine a situation where this method should be overrided
- the main reason to make this method virtual is to keep interface consistent

So, again, I suggest removing the _virtual_ specifier and reimplementing the method in a base class using other virtual methods.

**Comment by vpisarev** [comment]
@mshabunin, let me assign this one to you

**Comment by mshabunin** [comment]
@andreasBihlmaier , looks good, could you please squash commits into one?

**Comment by mshabunin** [comment]
Moved to #6103

**Pull request closed without merging**